substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs

```diff
@@ -1395,6 +1395,64 @@ where
 		}
 	}
 
+	/// Fetches the 'provides' tags associated to transactions part of blocks
+	/// of a retracted fork.
+	///
+	/// The 'provides' tags of transactions from retracted blocks are found
+	/// in inactive views. They can be used as a cache in the procces of transactions
+	/// prunning from the new best view in case of handling a new best block part of
+	/// a new fork. Usually the new best view can have 'provides' for part of the
+	/// transactions on the enacted fork, but the rest need to be revalidated, which
+	/// is costly, so using the inactive view's 'provides' tags for those transactions
+	/// saves time.
+	async fn provides_tags_from_inactive_views(
+		&self,
+		tree_route: &TreeRoute<Block>,
+	) -> HashMap<ExtrinsicHash<ChainApi>, Vec<Tag>> {
+		future::join_all(
+			tree_route
+				.retracted()
+				.iter()
+				.chain(std::iter::once(tree_route.common_block()))
+				.map(|hn| {
+					let api = self.api.clone();
+					async move {
+						let view = self.view_store.get_view_at(hn.hash, true);
+						let xts = api
+							.block_body(hn.hash)
+							.await
+							.unwrap_or_else(|e| {
+								log::warn!(target: LOG_TARGET, "provides_tags_from_inactive_views: block_body error request: {}", e);
+								None
+							})
+							.unwrap_or_default();
+						let xts_hashes = xts
+							.iter()
+							.filter_map(|xts| {
+								view.as_ref().map(|(inner, _)| inner.pool.hash_of(xts))
+							})
+							.collect::<Vec<_>>();
+
+						let xts_provides_tags = view
+							.map(|(inner, _)| {
+								inner.pool.validated_pool().extrinsics_tags(&xts_hashes)
+							})
+							.unwrap_or(vec![]);
+						xts_hashes
+							.into_iter()
+							.zip(xts_provides_tags.into_iter())
+							.into_iter()
+							.filter_map(|(hash, tags)| tags.map(|inner| (hash, inner)))
+							.collect::<Vec<(ExtrinsicHash<ChainApi>, Vec<Tag>)>>()
```

Comment:
Flow here needs to be changed.

The reason is that the view for block N will not contain any transactions from block N because they were already pruned.

So the proper flow is:
for every block N on retracted path, check all the views on retracted path from N-1 to common block.

Or alternatively (which is easier to reason about and also easier to implement) we could collect all transactions included in retracted blocks, and look for tags in the view_store (for the views that are on retracted path + common block). `ViewStore` shall have a new function for this.

For initial reference implementation you could just collect tags for all views in the view_store, and then optimize the searched set of views.

Reply:
> will ignore the txs provides tags that were used in building the common block

do we need them?

Reply:
> do we need them?

Ah, we don't look for the txs `provides` tags of the `common block`, as a matter of fact. The prune logic considers only the txs from the enacted fork (excluding the common block) to be considered when prunning the txs from the pool of the best new view. It isn't a problem, so please ignore.

Reply:
Based on our offline discussion, we should iterate through enacted blocks, get txs from there and search their `provides` tags in the inactive views corresponding to the retracted fork. The flow was updated accordingly here: [bbafba7](https://github.com/paritytech/polkadot-sdk/pull/7980/commits/bbafba78d13f79ede80c4aa02b0957a3235785dd) + here: [more polishing](https://github.com/paritytech/polkadot-sdk/pull/7980/commits/1f65b0afba9aabcac189038fa75fedcd195bbc6b).